src/server/pfs/server/commit_set.go

```diff
@@ -0,0 +1,455 @@
+func oldestAncestor(startCommit *pfs.CommitInfo, skipSet map[string]*pfs.CommitInfo) *pfs.Commit {
+	oldest := traverseToEdges(startCommit, skipSet, func(commitInfo *pfs.CommitInfo) []*pfs.Commit {
+		return []*pfs.Commit{commitInfo.ParentCommit}
+	})
+	if len(oldest) == 0 {
+		return nil
+	}
+	return oldest[0]
+}
+
+// traverseToEdges does a breadth first search using a traverse function.
+// returns all of the commits that can not continue traversal within the skip set, hence returning the known 'leaves' of the skipSet graph
+func traverseToEdges(startCommit *pfs.CommitInfo, skipSet map[string]*pfs.CommitInfo, traverse func(*pfs.CommitInfo) []*pfs.Commit) []*pfs.Commit {
+	cs := []*pfs.Commit{startCommit.Commit}
+	result := make([]*pfs.Commit, 0)
+	var c *pfs.Commit
+	for len(cs) > 0 {
+		c, cs = cs[0], cs[1:]
+		if c == nil {
+			continue
+		}
+		ci, ok := skipSet[pfsdb.CommitKey(c)]
+		if ok {
+			cs = append(cs, traverse(ci)...)
+		} else {
+			result = append(result, proto.Clone(c).(*pfs.Commit))
+		}
+	}
+	return result
+}
+
+// deleteCommits accepts commitInfos that may span commit sets.
+func (d *driver) deleteCommits(txnCtx *txncontext.TransactionContext, commitInfos []*pfs.CommitInfo) error {
+	deleteCommits := make(map[string]*pfs.CommitInfo)
+	repos := make(map[string]*pfs.Repo)
+	for _, ci := range commitInfos {
+		deleteCommits[pfsdb.CommitKey(ci.Commit)] = ci
+		repos[pfsdb.RepoKey(ci.Commit.Repo)] = ci.Commit.Repo
+	}
+	// delete the commits and their filesets
+	for _, ci := range commitInfos {
+		// make sure all children are finished, so we don't lose data
+		for _, child := range ci.ChildCommits {
+			if _, ok := deleteCommits[pfsdb.CommitKey(child)]; ok {
+				// this child is being deleted, any files from this commit will end up
+				// as part of *its* children, which have already been checked
+				continue
+			}
+			var childInfo pfs.CommitInfo
+			if err := d.commits.ReadWrite(txnCtx.SqlTx).Get(child, &childInfo); err != nil {
+				return errors.Wrapf(err, "error checking child commit state")
+			}
+			if childInfo.Finished == nil {
+				var suffix string
+				if childInfo.Finishing != nil {
+					// user might already have called "finish",
+					suffix = ", consider using WaitCommit"
+				}
+				return errors.Errorf("cannot squash until child commit %s is finished%s", child, suffix)
+			}
+		}
+		if err := d.commits.ReadWrite(txnCtx.SqlTx).Delete(ci.Commit); err != nil {
+			return errors.EnsureStack(err)
+		}
+		// Delete the commit's filesets
+		if err := d.commitStore.DropFileSetsTx(txnCtx.SqlTx, ci.Commit); err != nil {
+			return errors.EnsureStack(err)
+		}
+	}
+	// update branch heads
+	headlessBranches := make([]*pfs.BranchInfo, 0)
+	for _, repo := range repos {
+		repoInfo := &pfs.RepoInfo{}
+		if err := d.repos.ReadWrite(txnCtx.SqlTx).Get(repo, repoInfo); err != nil {
+			return err
+		}
+		branchInfo := &pfs.BranchInfo{}
+		for _, b := range repoInfo.Branches {
```

Comment:
This is rough in terms of scalability. We probably scale better with repos than branches.